File: src/dazzle_dnr_ui/specs/actions.py
"""
Action and effect specification types for UISpec.

Defines actions, effects, transitions, and state patches.
"""

from enum import Enum
from typing import Any, Literal

from pydantic import BaseModel, Field

from dazzle_dnr_ui.specs.state import Binding


# =============================================================================
# State Patches
# =============================================================================


class PatchOp(str, Enum):
    """Patch operations (JSON Patch style)."""

    SET = "set"
    DELETE = "delete"
    MERGE = "merge"
    APPEND = "append"
    REMOVE = "remove"


class PatchSpec(BaseModel):
    """
    State patch specification.

    Example:
        PatchSpec(op=PatchOp.SET, path="selectedClient", value=client_id)
        PatchSpec(op=PatchOp.APPEND, path="notifications", value=new_notification)
    """

    op: PatchOp = Field(description="Patch operation")
    path: str = Field(description="State path to patch")
    value: Any | None = Field(default=None, description="Value for the patch")

    class Config:
        frozen = True


# =============================================================================
# Transitions
# =============================================================================


class TransitionSpec(BaseModel):
    """
    State transition specification.

    Example:
        TransitionSpec(
            target_state="workspace.selectedClient",
            update=PatchSpec(op=PatchOp.SET, path="selectedClient", value=client)
        )
    """

    target_state: str = Field(description="Target state variable")
    update: PatchSpec = Field(description="Patch to apply")

    class Config:
        frozen = True


# =============================================================================
# Effects
# =============================================================================


class FetchEffect(BaseModel):
    """
    Fetch effect - call backend service.

    Example:
        FetchEffect(
            backend_service="list_clients",
            on_success="handleClientsLoaded",
            on_error="handleError"
        )
    """

    kind: Literal["fetch"] = "fetch"
    backend_service: str = Field(description="Backend service name to call")
    inputs: dict[str, Binding] | None = Field(
        default=None, description="Input bindings for the service"
    )
    on_success: str | None = Field(
        default=None, description="Action to dispatch on success"
    )
    on_error: str | None = Field(
        default=None, description="Action to dispatch on error"
    )

    class Config:
        frozen = True


class NavigateEffect(BaseModel):
    """
    Navigate effect - change route.

    Example:
        NavigateEffect(route="/clients/:id", params={"id": StateBinding(path="selectedClient.id")})
    """

    kind: Literal["navigate"] = "navigate"
    route: str = Field(description="Route path to navigate to")
    params: dict[str, Binding] | None = Field(
        default=None, description="Route parameters"
    )

    class Config:
        frozen = True


class LogEffect(BaseModel):
    """
    Log effect - console.log.

    Example:
        LogEffect(message=LiteralBinding(value="User clicked button"))
    """

    kind: Literal["log"] = "log"
    message: Binding = Field(description="Message to log")
    level: str = Field(default="info", description="Log level (info, warn, error)")

    class Config:
        frozen = True


class ToastEffect(BaseModel):
    """
    Toast effect - show notification.

    Example:
        ToastEffect(
            message=LiteralBinding(value="Client created successfully"),
            variant="success"
        )
    """

    kind: Literal["toast"] = "toast"
    message: Binding = Field(description="Toast message")
    variant: str = Field(
        default="info", description="Toast variant (info, success, warning, error)"
    )
    duration: int | None = Field(
        default=3000, description="Duration in milliseconds"
    )

    class Config:
        frozen = True


class CustomEffect(BaseModel):
    """
    Custom effect - extensibility point.

    Example:
        CustomEffect(name="analytics.track", config={"event": "button_click"})
    """

    kind: Literal["custom"] = "custom"
    name: str = Field(description="Custom effect name")
    config: dict[str, Any] = Field(
        default_factory=dict, description="Effect configuration"
    )

    class Config:
        frozen = True


# Union type for all effects
EffectSpec = FetchEffect | NavigateEffect | LogEffect | ToastEffect | CustomEffect


# =============================================================================
# Actions
# =============================================================================


class ActionSpec(BaseModel):
    """
    Action specification.

    Actions are dispatched in response to user interactions or effects.

    Example:
        ActionSpec(
            name="selectClient",
            inputs=SchemaSpec(fields=[SchemaFieldSpec(name="client_id", type="uuid")]),
            transitions=[
                TransitionSpec(
                    target_state="workspace.selectedClient",
                    update=PatchSpec(op=PatchOp.SET, path="selectedClient", value=client_id)
                )
            ],
            effect=NavigateEffect(route="/clients/:id", params={"id": client_id})
        )
    """

    name: str = Field(description="Action name")
    description: str | None = Field(default=None, description="Action description")
    inputs: dict[str, str] | None = Field(
        default=None, description="Input parameters (name -> type)"
    )
    transitions: list[TransitionSpec] = Field(
        default_factory=list, description="State transitions to apply"
    )
    effect: EffectSpec | None = Field(
        default=None, description="Side effect to execute"
    )

    class Config:
        frozen = True

File: src/dazzle_dnr_ui/runtime/static/js/actions.js
// @ts-check
/**
 * DNR-UI Actions & Effects - Action execution system
 * Part of the Dazzle Native Runtime
 *
 * @module actions
 */

import { batch } from './signals.js';
import { setState, updateState, setGlobalLoading, setGlobalError } from './state.js';
import { resolveBinding } from './bindings.js';
import { apiClient } from './api-client.js';
import { showToast } from './toast.js';

/** @type {((name: string, payload: any) => void) | null} */
let actionLogger = null;

/**
 * Set a function to log action dispatches (used by devtools).
 * @param {(name: string, payload: any) => void} logger
 */
export function setActionLogger(logger) {
  actionLogger = logger;
}

// =============================================================================
// Action Registry
// =============================================================================

const actionRegistry = new Map();

export function registerAction(name, handler) {
  actionRegistry.set(name, handler);
}

/**
 * Dispatch an action by name.
 * @param {string} actionName - Name of the action to dispatch
 * @param {Object} [payload={}] - Action payload
 * @returns {any}
 */
export function dispatch(actionName, payload = {}) {
  // Log to devtools if available
  if (actionLogger) {
    actionLogger(actionName, payload);
  }

  const handler = actionRegistry.get(actionName);
  if (handler) {
    return handler(payload);
  }
  console.warn(`Action not found: ${actionName}`);
}

export function hasAction(name) {
  return actionRegistry.has(name);
}

// =============================================================================
// Action Execution
// =============================================================================

export async function executeAction(action, context) {
  if (!action) return;

  const actionName = typeof action === 'string' ? action : action.name;

  // Check for registered actions first
  if (actionRegistry.has(actionName)) {
    return dispatch(actionName, context.payload || context.data || {});
  }

  // Apply state transitions with batch for performance
  if (action.transitions && action.transitions.length > 0) {
    batch(() => {
      action.transitions.forEach(transition => {
        const scope = transition.scope || 'local';
        const targetState = transition.target_state || transition.targetState;
        const path = scope === 'local' ?
          `${context.componentId}_${targetState}` : targetState;

        // Handle patch operations
        const update = transition.update;
        if (update && update.op) {
          applyPatch(scope, path, update, context);
        } else {
          // Simple set
          const value = resolveBinding(transition.value || update, context);
          setState(scope, path, value);
        }
      });
    });
  }

  // Execute effect
  if (action.effect) {
    await executeEffect(action.effect, context);
  }
}

// =============================================================================
// Patch Operations
// =============================================================================

export function applyPatch(scope, path, patch, context) {
  const value = resolveBinding(patch.value, context);

  switch (patch.op) {
    case 'set':
      setState(scope, path, value);
      break;

    case 'merge':
      updateState(scope, path, current => ({ ...current, ...value }));
      break;

    case 'append':
      updateState(scope, path, current => Array.isArray(current) ? [...current, value] : [value]);
      break;

    case 'remove':
      updateState(scope, path, current => {
        if (Array.isArray(current)) {
          return current.filter(item => {
            if (typeof value === 'function') return !value(item);
            if (typeof value === 'object' && value.id) return item.id !== value.id;
            return item !== value;
          });
        }
        return current;
      });
      break;

    case 'delete':
      setState(scope, path, undefined);
      break;
  }
}

// =============================================================================
// Effect Execution
// =============================================================================

export async function executeEffect(effect, context) {
  if (!effect) return;

  switch (effect.kind) {
    case 'fetch': {
      const service = effect.backend_service || effect.backendService;
      const method = effect.method || 'GET';
      const inputs = {};

      // Resolve input bindings
      Object.entries(effect.inputs || {}).forEach(([key, binding]) => {
        inputs[key] = resolveBinding(binding, context);
      });

      setGlobalLoading(true);
      setGlobalError(null);

      try {
        let result;
        const entity = service.replace(/_service$/, 's');

        // Map to appropriate API call based on method
        if (method === 'GET' || effect.operation === 'list') {
          result = await apiClient.list(entity, inputs);
        } else if (effect.operation === 'read' && inputs.id) {
          result = await apiClient.read(entity, inputs.id);
        } else if (method === 'POST' || effect.operation === 'create') {
          result = await apiClient.create(entity, inputs);
        } else if (method === 'PUT' || effect.operation === 'update') {
          result = await apiClient.update(entity, inputs.id, inputs);
        } else if (method === 'DELETE' || effect.operation === 'delete') {
          result = await apiClient.remove(entity, inputs.id);
        } else {
          // Generic fetch
          result = await apiClient.request(method, `/${service}`, inputs);
        }

        // Handle success
        if (effect.on_success || effect.onSuccess) {
          const successAction = context.getAction ?
            context.getAction(effect.on_success || effect.onSuccess) :
            { name: effect.on_success || effect.onSuccess };

          if (successAction) {
            await executeAction(successAction, { ...context, data: result, result });
          }
        }

        return result;
      } catch (error) {
        setGlobalError(error);
        console.error('Fetch effect failed:', error);

        // Handle error
        if (effect.on_error || effect.onError) {
          const errorAction = context.getAction ?
            context.getAction(effect.on_error || effect.onError) :
            { name: effect.on_error || effect.onError };

          if (errorAction) {
            await executeAction(errorAction, { ...context, error });
          }
        } else {
          // Default error handling
          showToast(error.message || 'An error occurred', { variant: 'error' });
        }
      } finally {
        setGlobalLoading(false);
      }
      break;
    }

    case 'navigate': {
      const route = effect.route;
      const params = {};
      Object.entries(effect.params || {}).forEach(([key, binding]) => {
        params[key] = resolveBinding(binding, context);
      });

      // Replace route params
      let url = route;
      Object.entries(params).forEach(([key, value]) => {
        url = url.replace(`{${key}}`, encodeURIComponent(value));
        url = url.replace(`:${key}`, encodeURIComponent(value));
      });

      window.history.pushState({ params }, '', url);
      window.dispatchEvent(new CustomEvent('dnr-navigate', { detail: { url, params } }));
      break;
    }

    case 'log': {
      const message = resolveBinding(effect.message, context);
      const level = effect.level || 'info';
      console[level](message);
      break;
    }

    case 'toast': {
      const message = resolveBinding(effect.message, context);
      showToast(message, {
        variant: effect.variant || 'info',
        duration: effect.duration || 3000
      });
      break;
    }

    case 'custom': {
      const handler = actionRegistry.get(`effect:${effect.name}`);
      if (handler) {
        await handler(effect.config, context);
      } else {
        console.warn(`Custom effect not found: ${effect.name}`);
      }
      break;
    }
  }
}

// =============================================================================
// Built-in Actions
// =============================================================================

// Pure actions (state only, no side effects)
registerAction('filter', ({ items, predicate, target }) => {
  if (!items || !predicate) return items;
  const filtered = items.filter(predicate);
  if (target) setState('workspace', target, filtered);
  return filtered;
});

registerAction('sort', ({ items, key, direction = 'asc', target }) => {
  if (!items || !key) return items;
  const sorted = [...items].sort((a, b) => {
    const aVal = a[key];
    const bVal = b[key];
    const cmp = aVal < bVal ? -1 : aVal > bVal ? 1 : 0;
    return direction === 'desc' ? -cmp : cmp;
  });
  if (target) setState('workspace', target, sorted);
  return sorted;
});

registerAction('select', ({ item, target }) => {
  if (target) setState('workspace', target, item);
  return item;
});

registerAction('toggle', ({ path, scope = 'workspace' }) => {
  updateState(scope, path, current => !current);
});

registerAction('reset', ({ path, scope = 'workspace', initial = null }) => {
  setState(scope, path, initial);
});

export { actionRegistry };
